core/src/main/java/org/geysermc/geyser/session/cache/tags/HolderSet.java

```diff
@@ -0,0 +1,114 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.geyser.session.cache.tags;
+
+import java.util.List;
+import java.util.function.Function;
+import lombok.Data;
+import net.kyori.adventure.key.Key;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.geysermc.geyser.session.GeyserSession;
+import org.geysermc.geyser.session.cache.TagCache;
+import org.geysermc.geyser.session.cache.registry.JavaRegistryKey;
+
+/**
+ * Similar to vanilla Minecraft's HolderSets, stores either a tag or a list of IDs (this list can also be represented as a single ID in vanilla HolderSets).
+ *
+ * <p>Because HolderSets utilise tags, when loading a HolderSet, Geyser must store tags for the registry the HolderSet is for (see {@link JavaRegistryKey}).</p>
+ *
+ * <p>Use the {@link HolderSet#readHolderSet} method to easily read a HolderSet from NBT sent by a server. To turn the HolderSet into a list of network IDs, use the {@link HolderSet#resolveRaw} method.
+ * To turn the HolderSet into a list of objects, use the {@link HolderSet#resolve} method.</p>
+ */
+@Data
+public final class HolderSet<T> {
+
+    private final JavaRegistryKey<T> registry;
+    private final @Nullable Tag<T> tag;
+    private final int @Nullable [] holders;
+
+    public HolderSet(JavaRegistryKey<T> registry, int @NonNull [] holders) {
+        this.registry = registry;
+        this.tag = null;
+        this.holders = holders;
+    }
+
+    public HolderSet(JavaRegistryKey<T> registry, @NonNull Tag<T> tagId) {
+        this.registry = registry;
+        this.tag = tagId;
+        this.holders = null;
+    }
+
+    /**
+     * Resolves the HolderSet, and automatically maps the network IDs to their respective object types. If the HolderSet is a list of IDs, this will be returned. If it is a tag, the tag will be resolved from the tag cache.
+     *
+     * @return the HolderSet turned into a list of objects.
+     */
+    public List<T> resolve(GeyserSession session) {
+        return TagCache.mapRawArray(session, resolveRaw(session.getTagCache()), registry);
+    }
+
+    /**
+     * Resolves the HolderSet. If the HolderSet is a list of IDs, this will be returned. If it is a tag, the tag will be resolved from the tag cache.
+     *
+     * @return the HolderSet turned into a list of objects.
+     */
+    public int[] resolveRaw(TagCache tagCache) {
+        if (holders != null) {
+            return holders;
+        }
+
+        assert tag != null;
```

Comment:
Gotta be careful with asserts since they're a Throwable and not an Exception; we can replace this with Objects.requireNonNull if we want a check here.